packages/obligatron/src/obligations/tagging.ts

```
@@ -1,68 +1,103 @@
-			arn,
-			service,
-			resource_type,
-			bool_or(taggable) as taggable,
-			jsonb_aggregate(tags) as tags
-		FROM aws_resources_raw()
-		WHERE taggable = true
-		GROUP BY account_id, arn, service, resource_type;
-  `;
+	const findings = await db.aws_securityhub_findings.findMany({
+		where: {
+			product_fields: {
+				path: ['StandardsArn'],
+				string_starts_with:
+					'arn:aws:securityhub:::standards/aws-resource-tagging-standard',
+			},
+			compliance: {
+				path: ['Status'],
+				equals: 'FAILED',
+			},
+		},
+	});
+
+	console.log({
+		message: 'Received findings from security hub',
+		total: findings.length,
+	});
 
 	const results: ObligationResult[] = [];
 
-	for (const resource of awsResources) {
-		if (isExemptResource(resource)) {
+	for (const finding of findings) {
+		const resources = finding.resources?.valueOf();
+
+		if (!Array.isArray(resources)) {
+			console.error({
+				message: `Skipping invalid SecurityHub finding, invalid 'resources' field`,
+				finding_id: finding.id,
+			});
 			continue;
 		}
 
-		for (const requiredTag of REQUIRED_TAGS) {
-			if (resourceHasTag(resource, requiredTag)) {
-				continue;
+		// The Security Hub spec indicates that a finding can have multiple resources,
+		// I don't think this will happen for the Tagging rules, but lets be safe by
+		// handling this situation and raising a warning.
+		if (resources.length !== 1) {
+			console.warn({
+				message: `Finding had more (or less) that 1 resource: ${resources.length}`,
+				finding_id: finding.id,
+			});
+		}
+
+		for (const [index, resource] of resources.entries()) {
+			// This in theory should not happen as long as AWS don't change their schema.
+			// if they do change the schema its unlikely to be just this one finding failing,
+			// so lets make sure that we crash the lambda and get a humans attention!
+			if (!isFindingResource(resource)) {
+				throw new Error(
+					`Invalid resource in finding ${finding.id} at index ${index}`,
+				);
```

IIUC this `Error` could halt execution at any point in the for loop. Is this desired? Is there a scenario where in a list of 10 items, only index 5 is faulty? Throwing in this way means 6-10 don't get processed?

I might have mis-understood the code too!

---

Yep this is intentional, it will kill the entire lambda not just the 1 loop. Theres a comment just above explaining my reasoning!

TLDR: in theory this should only fail if AWS change their schema and in that case I want the lambda to fail fast and alert us.

---

I think we might be able to fail faster by doing this check up front, either within the SQL query itself, or as a `filter` on the dataset before looping over each item.

These changes will only have a performance impact over a certain scale. I wonder if throwing outside of the loop improves readability, and maintainability?

Not a blocker though.